File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
    bool solve(vector<vector<int>>& dp, vector<int>& nums, int sz, int target, int currSum, int ind){
        if(target == currSum) return true;
        if(ind == sz) return false;

        if(dp[ind][currSum] != -1) return dp[ind][currSum];

        bool take = false, notTake = solve(dp, nums, sz, target, currSum, ind+1);
        if(nums[ind] + currSum <= target){
            take = solve(dp, nums, sz, target, currSum + nums[ind], ind+1);
        }
        return dp[ind][currSum] = take || notTake;
    }
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(), nums.end(), 0), sz = nums.size();
        vector<vector<int>> dp(sz, vector<int>(sum, -1));
        
        if(sum&1) return false;
        return solve(dp, nums, sz, sum/2, 0, 0);
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    private boolean ans = false;
    private int[][] vis;
    public boolean canPartition(int[] nums) {
        int sz = nums.length, sum = 0;
        vis = new int[sz][20001];

        for(int i=0; i<sz; i++){
            sum += nums[i];
        }

        if((sum&1) == 1) return false;

        solve(nums, sz, sum/2, 0);

        return ans;
    }

    private void solve(int[] nums, int sz, int sum, int ind){
        if(sum == 0){
            ans = true;
            return;
        }
        if(ind == sz || ans || vis[ind][sum] == 1) return;

        vis[ind][sum] = 1;
        solve(nums, sz, sum, ind+1);
        if(nums[ind] <= sum) solve(nums, sz, sum-nums[ind], ind+1);
    }
}